sycl/doc/design/MappingHostAddressesToDeviceEntities.md

```diff
@@ -0,0 +1,240 @@
+- `Kind` is a string which corresponds to `kind` argument passed to
+  `[[__sycl_detail__::uniquely_identifiable_object(kind)]]` attribute attached
+  to the type of the variable identified by `Address`. It can be used to
+  distinguish different entities like `specialization_id` and `device_global`:
+  for example they could be stored in different maps to speed up certain
+  operations with them.
+
+The compiler guarantees that the function will be called zero or more times
+(depending on the amount of uniquely identifiable objects found in a program)
+_before_ application's `main()` function, i.e. in a global constructor.
+
+That poses some restrictions on those uniquely identifiable object, i.e. that
+they can't be used from another global object due to risk of accessing a
+non-initialized object, but that is an UB anyway because the order of global
+objects initialization is not defined in C++ when those objects are defined in
+different translation unit.
+
+## Compiler driver part
+
+The compiler driver is the component which is responsible for selecting the
+approach we are taking and the decision is made based on whether or not
+3rd-party host compiler is in use.
+
+If `-fsycl-host-compiler` option is present, the compiler driver chooses the
+integration footer approach:
+- it supplies device compilation step with `-fsycl-int-footer` option to
+  instruct device compiler to emit integration footer
+- it appends the integration footer to user-provided translation unit before
+  passing it to a host compiler
+
+Otherwise, if `-fsycl-host-compiler` is not present, then the compiler driver
+chooses another approach by simply doing nothing related to integration footer:
+- `-fsycl-int-footer` is **not** passed to device compiler
+- user-provided translation unit is passes as-is to host compiler
+
+## Integration footer approach
+
+When this approach is used, not only extra file (integration footer) is
+generated, but integration header is also modified: FE compiler generates a
+definition of a namespace scope variable of type
+`__sycl_device_global_registration` whose sole purpose it to run its constructor
+before the application's `main()` function:
+
+```
+namespace sycl::detail {
+namespace {
+
+class __sycl_device_global_registration {
+ public:
+  __sycl_device_global_registration() noexcept;
+};
+__sycl_device_global_registration __sycl_device_global_registrar;
+
+} // namespace (unnamed)
+} // namespace sycl::detail
+```
+
+The integration footer generated by the compiler contains the definition of the
+constructor, which calls a function in the DPC++ runtime, which registers
+needed mappings:
+
+```
+namespace sycl::detail {
+namespace {
+
+__sycl_device_global_registration::__sycl_device_global_registration() noexcept {
+  __register_uniquely_identifiable_object(
+    &::Foo,
+    /* same string returned from __builtin_sycl_unique_stable_id(::Foo) */,
+    "specialization_id");
```

Comment:
What is the third parameter ("specialization_id") stands for?

Reply:
The third parameter should contain the string passed to `[[__sycl_detail__::uniquely_identifiable_object(kind)]]` attribute. There is a description for it in the section about common runtime part.